apex/contrib/optimizers/distributed_fused_adam.py

```diff
@@ -648,95 +647,152 @@ def grad_sync(self):
                 group=self.distributed_process_group,
             )
-            grad_norm_sq -= non_parallel_grad_norm_sq
+            self._grad_norm = grad_norm_sq.sqrt()
+        return self._grad_norm.detach()
 
-        return grad_norm_sq.sqrt()
+    def clip_grad_norm(self, max_norm, parameters=[], norm_type=2.0):
+        """Clips gradient norm of parameters in optimizer
 
-    def step(self, closure=None, scale=1.):
+        The norm is computed over all gradients together, as if they
+        were concatenated into a single vector. The scaling is
+        deferred until the optimizer step, which should be called
+        immediately after this function.
+
+        The computed grad norm is cached to avoid redundant
+        communication.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            parameters (iterable, optional): an iterable of parameters
+                in optimizer (default: all parameters in optimizer).
+            norm_type (float or int, optional): type of the used
+                p-norm (default: 2)
+
+        """
+        assert max_norm > 0
+        total_norm = self.grad_norm(parameters=parameters, norm_type=norm_type)
+        inv_clip_coef = (total_norm + 1e-6) / max_norm
+        self._inv_grad_scale = torch.clamp(inv_clip_coef, min=1.0).view(1)
+        return total_norm
+
+    def step(self, grad_scaler=None):
         """Apply Adam optimizer step
 
         Arguments:
-            closure (callable, optional): closure to recompute loss
-                (default: None)
-            scale (float, optional): scaling factor to divide
-                gradients (default: 1.0)
+            grad_scaler (torch.cuda.amp.GradScaler, optional):
+                gradient scaler (default: None)
 
         """
-        self.state['step'] += 1
-        loss = None
-        if closure is not None:
-            loss = closure()
 
         # Make sure that gradients have been reduced
         self.grad_sync()
 
-        # Scale gradient if L2 norm is too large
-        if self.max_grad_norm > 0:
-            grad_norm = self.grad_norm().item()
-            if (math.isfinite(grad_norm)
-                and grad_norm / scale > self.max_grad_norm):
-                scale = grad_norm / self.max_grad_norm
+        # Apply gradient scaler if provided
+        # Note: We compute gradient norm to check for non-finite
+        # values. This is more conservative and compute intensive than
+        # directly checking, but it avoids extra communication if we
+        # have already computed gradient norm e.g. for gradient
+        # clipping.
+        if grad_scaler is not None:
+            grad_norm = self.grad_norm()
+            found_inf = torch.logical_not(torch.isfinite(grad_norm))
+            scaler_state = grad_scaler._per_optimizer_states[id(self)]
+            scaler_state['found_inf_per_device'] = {found_inf.device: found_inf.float()}
```

Comment:
This cast here is redundant if you assume that `grad_scaler` is an instance of `torch.cuda.amp.GradScaler`
```suggestion
            scaler_state['found_inf_per_device'] = {found_inf.device: found_inf}
```

Reply:
I don't think `GradScaler` does any casting. I got runtime errors when I passed in a `bool`.

Reply:
uh this logic overloads `torch.cuda.amp.GradScaler`'s `found_inf` logic. I see